core/src/main/java/org/infinispan/util/ByteString.java

```diff
@@ -0,0 +1,90 @@
+package org.infinispan.util;
+
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.marshall.core.Ids;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * @author Tristan Tarrant
+ * @since 9.0
+ */
+public class ByteString {
```

Comment:
Different encoding/hashing...etc can be achieved with a static class that gets the byte[] and produces what you want. I'm not 100% sure we really need a wrapper here, but feel free to go ahead with it, we can always change later.

The only real problem we have with byte[] is equality and hashcode but that's already sorted out with the equivalence functions that are separated out.